Bug: Multi-return type validation does not check individual types

## Description

When assigning multi-return function results to variables, the type checker does not validate that each individual return type matches the corresponding variable type. This allows type-reversed assignments to compile and run.

## Reproduction

```ez
import @std
using std

do get_pair() -> (int, string) {
    return 42, "hello"
}

do main() {
    // Types are reversed - should FAIL at compile time
    temp x string, y int = get_pair()
    println("x=${x}, y=${y}")  // Prints: x=42, y=hello
}
```

## Expected Behavior

The type checker should emit an error like:
```
error[E3001]: type mismatch in multi-return assignment: expected string for 'x', got int
error[E3001]: type mismatch in multi-return assignment: expected int for 'y', got string
```

## Actual Behavior

The code compiles without error and runs, assigning the int `42` to string variable `x` and string `"hello"` to int variable `y`.

## Impact

High - completely bypasses type safety for multi-return functions.

--- a/integration-tests/fail/errors/E3001_multi_return_type_mismatch.ez
+++ b/integration-tests/fail/errors/E3001_multi_return_type_mismatch.ez
@@ -1,0 +1,13 @@
+/*
+ * Error Test: E3001 - multi-return-type-mismatch
+ * Expected: "type mismatch" or "multi-return"
+ */
+
+do get_pair() -> (int, string) {
+    return 42, "hello"
+}
+
+do main() {
+    // Types reversed - should fail
+    temp x string, y int = get_pair()
+}

--- a/pkg/typechecker/typechecker.go
+++ b/pkg/typechecker/typechecker.go
@@ -1033,9 +1033,7 @@ func (tc *TypeChecker) checkStatement(stmt ast.Statement, expectedReturnTypes []
 func (tc *TypeChecker) checkVariableDeclaration(decl *ast.VariableDeclaration) {
 	// Handle multiple names (for multi-return assignment)
 	if len(decl.Names) > 1 {
-		// Multi-return assignment like: temp result, err = divide(10, 2)
-		// For now, we can't easily infer types for multi-return
-		// The runtime will handle this
+		tc.checkMultiReturnDeclaration(decl)
 		return
 	}
 
@@ -1249,6 +1247,84 @@ func (tc *TypeChecker) checkVariableDeclaration(decl *ast.VariableDeclaration) {
 	// Register variable in current scope with mutability (temp = mutable, const = immutable)
 	if declaredType != "" {
 		tc.defineVariableWithMutability(varName, declaredType, decl.Mutable)
+	}
+}
+
+// checkMultiReturnDeclaration validates multi-return variable declarations
+// e.g., temp x int, y string = getValues()
+func (tc *TypeChecker) checkMultiReturnDeclaration(decl *ast.VariableDeclaration) {
+	// Check for type/function used as value
+	if decl.Value != nil {
+		tc.checkValueExpression(decl.Value)
+		tc.checkExpression(decl.Value)
+	}
+
+	// If no explicit types declared, nothing to check at compile time
+	if len(decl.TypeNames) == 0 {
+		return
+	}
+
+	// Get the function call to check return types
+	callExpr, ok := decl.Value.(*ast.CallExpression)
+	if !ok {
+		// Value is not a function call, can't check return types
+		return
+	}
+
+	// Get the function name
+	var funcName string
+	switch fn := callExpr.Function.(type) {
+	case *ast.Label:
+		funcName = fn.Value
+	case *ast.MemberExpression:
+		// Module.function call - get the function part
+		funcName = fn.Member.Value
+	default:
+		return
+	}
+
+	// Look up the function signature
+	funcSig, exists := tc.functions[funcName]
+	if !exists {
+		// Function not found - will be caught elsewhere
+		return
+	}
+
+	// Check that the number of declared types matches the number of return types
+	if len(decl.TypeNames) != len(funcSig.ReturnTypes) {
+		// Count mismatch will be caught by E5012 at runtime
+		return
+	}
+
+	// Check each declared type against the corresponding return type
+	for i, declaredType := range decl.TypeNames {
+		if i >= len(funcSig.ReturnTypes) {
+			break
+		}
+		returnType := funcSig.ReturnTypes[i]
+
+		if !tc.typesCompatible(declaredType, returnType) {
+			// Get position from the variable name at this index
+			line, col := 0, 0
+			if i < len(decl.Names) && decl.Names[i] != nil {
+				line = decl.Names[i].Token.Line
+				col = decl.Names[i].Token.Column
+			}
+			tc.addError(
+				errors.E3001,
+				fmt.Sprintf("type mismatch in multi-return: variable '%s' declared as %s but function returns %s",
+					decl.Names[i].Value, declaredType, returnType),
+				line,
+				col,
+			)
+		}
+	}
+
+	// Register variables in scope
+	for i, name := range decl.Names {
+		if name != nil && i < len(decl.TypeNames) {
+			tc.defineVariableWithMutability(name.Value, decl.TypeNames[i], decl.Mutable)
+		}
 	}
 }
 

